statsmodels/base/_penalties.py

```diff
@@ -73,54 +86,457 @@ def grad(self, params):
+        c0 = self.c0
+        deriv_c0 = super(SCADSmoothed, self).grad(c0)
+        value_c0 = super(SCADSmoothed, self).func(c0)
+        self.aq1 = value_c0 - 0.5 * deriv_c0 * c0
+        self.aq2 = 0.5 * deriv_c0 / c0
+        self.restriction = restriction
+
+    def func(self, params):
+        # TODO: `and np.size(params) > 1` is hack for llnull, need better solution
+        if self.restriction is not None and np.size(params) > 1:
+            params = self.restriction.dot(params)
+        # need to temporarily override weights for call to super
+        # Note: we have the same problem with `restriction`
+        weights = self.weights
+        self.weights = 1.
+        value = super(SCADSmoothed, self).func(params[None, ...])
+        self.weights = weights
+
+        # shift down so func(0) == 0
+        value -= self.aq1
+        #change the segment corrsponding to quadratic approximation
+        p_abs = np.atleast_1d(np.abs(params))
+        mask = p_abs < self.c0
+        p_abs_masked = p_abs[mask]
+        value[mask] = self.aq2 * p_abs_masked**2
+
+        return (self.weights * value).sum(0)
+
+    def grad(self, params):
+        if self.restriction is not None and np.size(params) > 1:
+            params = self.restriction.dot(params)
+        # need to temporarily override weights for call to super
+        weights = self.weights
+        self.weights = 1.
+        value = super(SCADSmoothed, self).grad(params)
+        self.weights = weights
+
+        #change the segment corrsponding to quadratic approximation
+        p = np.atleast_1d(params)
+        mask = np.abs(p) < self.c0
+        value[mask] = 2 * self.aq2 * p[mask]
+
+        if self.restriction is not None and np.size(params) > 1:
+            return weights * value.dot(self.restriction)
+        else:
+            return weights * value
+
+    def deriv2(self, params):
+        if self.restriction is not None and np.size(params) > 1:
+            params = self.restriction.dot(params)
+        # need to temporarily override weights for call to super
+        weights = self.weights
+        self.weights = 1.
+        value = super(SCADSmoothed, self).deriv2(params)
+        self.weights = weights
+
+        # change the segment corrsponding to quadratic approximation
+        p = np.atleast_1d(params)
+        mask = np.abs(p) < self.c0
+        value[mask] = 2 * self.aq2
+
+        if self.restriction is not None and np.size(params) > 1:
+            # note: super returns 1d array for diag, i.e. hessian_diag
+            # TODO: weights are missing
+            return (self.restriction.T * value).dot(self.restriction)
+        else:
+            return value
+
+
+class ConstraintsPenalty(object):
+    """
+    Penalty applied to linear transformation of parameters
+
+    Parameters
+    ----------
+    penalty: instance of penalty function
+        currently this requires an instance of a univariate, vectorized
+        penalty class
```

Comment:
Should there be a check for this in `__init__` to fail early+loud?

Reply:
I'm not adding checks yet because, I still need to figure out whether and how it works with other classes. 
API is not final yet.